apps/api-extractor/src/generators/DeclarationReferenceGenerator.ts

```diff
@@ -99,68 +87,48 @@ export class DeclarationReferenceGenerator {
     );
   }
 
-  private static _getNavigationToSymbol(symbol: ts.Symbol): Navigation | 'global' {
+  private _getNavigationToSymbol(symbol: ts.Symbol): Navigation {
```

Comment:
Here's an interesting canonical reference question: Consider the following example:

```
// index.ts (entry point of "my package")
import * as n from './internal';
export function someFunction(): n.SomeType { return 5; }

// internal.ts
export type SomeType = number;
```

What should the canonical reference of `SomeType` be? Today...

* `DeclarationReferenceGenerator` produces the following: `my-package!~SomeType:type` (after this PR)
* whereas api-extractor-model produces the following: `my-package!~n.SomeType`

api-extractor-model produces the latter because the "synthetic" `AstNamespaceImport` for `n` makes its way into the `.api.json`.

I feel like the correct canonical reference in this case is `my-package!~SomeType:type`, because the namespace `n` is really an implementation detail of how `index.ts` imports from `internal.ts`, and not part of `SomeType` itself. There could be the following code in `internal.ts`:

```
export function someOtherFunction(): SomeType { return 6; }
```

and it would be odd for `someOtherFunction`'s reference token for `SomeType` to include the synthetic namespace `n`. I'd expect for it to be `my-package!~SomeType:type`.

Regardless, I don't think we should solve this in this PR, just something to think about.

Reply:
Hmm.... A central principle of API Extractor is that it ***interprets*** declarations, imposing a semantics that is not technically present in the source code.  This is what justifies rolling up declarations into a different file from where they were declared, discarding some .d.ts files entirely (because they aren't reachable from the entry point), and renaming API items to avoid naming conflicts.

Your question is really about what semantics should be applied to an unexported declaration.

1. I'm a little sketchy on [the grammar](https://github.com/microsoft/tsdoc/blob/main/tsdoc/src/beta/DeclarationReference.grammarkdown), but I think the physical location would be described as `my-package/lib/internal!~SomeType:type`.

2. But generally API Extractor's model is that of the .d.ts rollup. Physically in the .d.ts rollup, the location would be `my-package!~SomeType:type`.

3. However if there are naming collisions, this could become something arbitrary like `my-package!~SomeType_3:type`.

4. Logically, the human intent of `import * as n` is to move this declaration inside `n` which would better be described as `my-package!n~SomeType:type`.  The reason it does not get placed there was technical: if they also did `import * as n2`  then its ambiguous which one is the "real" definition and which one is the alias.  With more sophisticated rollup logic, we could solve this problem in the future, and then `my-package!n~SomeType:type` maybe would be correct. 

I guess the choice boils down to requirements:  

Is our priority that the canonical reference should help a tool to **find the declaration in the .d.ts files?**  If so then it should probably be a physical location in the .d.ts rollup.  If the .d.ts rollup feature is disabled, then maybe all canonical references should actually be physical locations such as (1).

Alternatively, is our priority that the canonical reference should be a **stable identifier**, i.e. resistant to getting shuffled around by trivial changes?  I think this is the design choice that API Extractor has pursued. In this case I guess we would go with (2) or (3).